csrc/scheduler/matmul_utils.cpp

```diff
@@ -314,144 +314,302 @@ std::string isMatmulFusionDefinitionSupported(
-  NVF_ERROR((int64_t)sizes.size() == tv->nDims());
-  size_t inner_dim_pos = 0;
-  for (size_t i = tv->getMaybeAllocationDomain().size() - 1; i >= 0; --i) {
-    IterDomain* id = tv->getMaybeAllocationDomain()[i];
-    if (id->isReduction() || id->isBroadcast()) {
-      continue;
-    }
-    inner_dim_pos = i;
-    std::optional<bool> c = tv->getContiguity().at(i);
-    NVF_ERROR(c.has_value());
-    if (!c.value()) {
-      // If TensorView is marked discontiguous in inner dimension, we cannot
-      // vectorize regardless of input.
-      return 1l;
-    } else {
-      NVF_CHECK(
-          strides[i] == 1,
-          "TensorView ",
-          tv->toString(),
-          " has marked contiguous inner dimension ",
-          id->toString(),
-          " but provided tensor has stride ",
-          strides[i],
-          " in that dimension.");
-    }
-    break; // only check innermost realized dimension
+  MatmulParams::SupportedVectorization compute() {
+    const std::vector<int64_t> op_vecs = operandVectorizations();
+    NVF_ERROR(op_vecs.size() == 2, "Expected exactly two operands");
+    return {op_vecs[0], op_vecs[1], epilogueVectorization()};
+  }
+
+ private:
+  std::vector<int64_t> operandVectorizations() {
+    std::vector<int64_t> vec_sizes;
+    for (MatmulRole role : {MatmulRole::INPUT_A, MatmulRole::INPUT_B}) {
+      const auto op_it = tensor_roles_.find(role);
+      if (op_it == tensor_roles_.end()) {
+        continue;
+      }
+      for (TensorView* tv : op_it->second) {
+        vec_sizes.push_back(operandVectorization(tv));
+      }
+    }
+    return vec_sizes;
   }
 
-  // Since this is unpredicated vectorization, the size of the innermost
-  // dimension must be a multiple of the vectorization factor.
-  vec_size = std::min(
-      vec_size, scheduler_utils::maxVectorizationWidth(sizes[inner_dim_pos]));
+  MatmulDomain dimRole(const ValGroup& g) const {
+    auto dim_role_it = dim_roles_.find(g);
+    NVF_ERROR(
+        dim_role_it != dim_roles_.end(), "Found ValGroup with unknown role");
+    return dim_role_it->second;
+  }
+
+  int64_t ptrAndDTypeVec(TensorView* tv) const {
+    const int64_t data_ptr_int = (int64_t)runtime_info_.ptrOf(tv);
+    int64_t vec_size = scheduler_utils::maxVectorizationWidth(data_ptr_int);
+    vec_size = std::min(vec_size, 16l);
+    vec_size /= dataTypeSize(tv->dtype());
+    vec_size = std::max(vec_size, 1l);
+    return vec_size;
+  }
+
+  // Note this is non-const because we use runtime_info_.expressionEvaluator()
+  std::pair<std::vector<int64_t>, std::vector<int64_t>> getSizesAndStrides(
+      TensorView* tv) {
+    if (tv->isFusionInput()) {
+      return {
+          runtime_info_.getInputAllocationSizes(tv),
+          runtime_info_.getInputAllocationStrides(tv)};
+    }
+    // For non-inputs, compute sizes using ExpressionEvaluator, then compute
+    // strides based on allocation domain, assuming full contiguity regardless
+    // of how it is marked in the TensorView.
```

Comment:
Could you explain more (in the code as comment) why assuming full contiguity?

Reply:
I changed this to respect the contiguity flags. Originally, I was mimicking how `FusionExecutor` allocates outputs: as fully contiguous regardless of these flags. However, we are now actually able to pass preallocated outputs (soon to be fully plumbed in, see #2247). The tensors we pass should be compatible with the allocation domain and contiguity marked in the `TensorView`, but are otherwise unconstrained. I updated the code to choose odd strides for discontiguous dimensions in case of preallocated outputs. We don't have a way to inspect the pointer alignment of such preallocated outputs but I put another comment in the code that this should be checked.